File: src/SoliBee.Core/Models/BlackjackState.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SoliBee.Core.Models;

public enum BlackjackPhase { Betting, Playing, DealerTurn, Result }

public enum BlackjackHandResult { Pending, Won, Lost, Push, Blackjack }

public class BlackjackHand
{
    public List<Card> Cards { get; set; } = new();
    public int Bet { get; set; } = 1;
    public bool IsDoubled { get; set; }
    public bool IsStood { get; set; }
    public bool FromSplit { get; set; }
    public BlackjackHandResult Result { get; set; } = BlackjackHandResult.Pending;

    // Full value (all cards, regardless of IsFaceUp) — used for game logic
    public (int Value, bool IsSoft) ComputeValue()
    {
        int value = 0, aces = 0;
        foreach (var c in Cards)
        {
            if (c.Rank == 1) { aces++; value += 11; }
            else value += Math.Min(c.Rank, 10);
        }
        while (value > 21 && aces > 0) { value -= 10; aces--; }
        return (value, aces > 0);
    }

    // Face-up cards only — used for display during Playing phase
    public (int Value, bool IsSoft) ComputeVisibleValue()
    {
        int value = 0, aces = 0;
        foreach (var c in Cards.Where(c => c.IsFaceUp))
        {
            if (c.Rank == 1) { aces++; value += 11; }
            else value += Math.Min(c.Rank, 10);
        }
        while (value > 21 && aces > 0) { value -= 10; aces--; }
        return (value, aces > 0);
    }

    public bool IsBust       => ComputeValue().Value > 21;
    public bool IsBlackjack  => !FromSplit && Cards.Count == 2 && ComputeValue().Value == 21;
    public bool IsComplete   => IsStood || IsBust || IsDoubled || IsBlackjack || ComputeValue().Value == 21;
}

public class BlackjackState
{
    public BlackjackPhase      Phase           { get; set; } = BlackjackPhase.Betting;
    public List<BlackjackHand> PlayerHands     { get; set; } = new() { new() };
    public BlackjackHand       DealerHand      { get; set; } = new();
    public int                 ActiveHandIndex { get; set; } = 0;
    public int                 Credits         { get; set; } = 100;
    public int                 CurrentBet      { get; set; } = 1;
    public bool                IsSplit         { get; set; }
    public int                 LastNetResult   { get; set; }
}

File: src/Blackjack/Models/BlackjackState.swift
import Foundation

public enum BlackjackPhase: Equatable {
    case betting
    case playing
    case dealerTurn
    case result
}

public enum BlackjackHandResult: Equatable {
    case win
    case loss
    case push
    case blackjack
    case bust
}

public struct BlackjackHand: Equatable {
    public var cards: [Card]
    public var bet: Int
    public var isDoubled: Bool = false
    public var isSplitAce: Bool = false
    public var result: BlackjackHandResult? = nil

    public init(cards: [Card], bet: Int) {
        self.cards = cards
        self.bet = bet
    }

    public var value: Int { BlackjackState.handValue(cards) }
    public var isBust: Bool { value > 21 }
    public var isBlackjack: Bool { cards.count == 2 && value == 21 }
}

public struct BlackjackState: Equatable {
    public var phase: BlackjackPhase = .betting
    public var playerHands: [BlackjackHand] = []
    public var activeHandIndex: Int = 0
    public var dealerCards: [Card] = []
    public var deck: [Card] = []
    public var sessionCredits: Int = 100
    public var currentBet: Int = 1
    public var handsDealt: Int = 0
    public var lastResultSummary: String = ""
    public var lastNetResult: Int = 0
    public var dealerValue: Int { BlackjackState.handValue(dealerCards) }
    public var dealerVisibleValue: Int {
        // Only count face-up cards during playing phase
        let visible = dealerCards.filter { $0.faceUp }
        return BlackjackState.handValue(visible)
    }

    // Compute best blackjack hand value (aces as 11 or 1)
    public static func handValue(_ cards: [Card]) -> Int {
        var total = 0
        var aces = 0
        for card in cards where card.faceUp {
            let r = card.rank
            if r == 1 { aces += 1; total += 11 }
            else if r >= 10 { total += 10 }
            else { total += r }
        }
        while total > 21 && aces > 0 { total -= 10; aces -= 1 }
        return total
    }
}
